plugin-server/src/main/ingestion-queues/analytics-events-ingestion-consumer.ts

```diff
@@ -78,50 +78,63 @@ export async function eachBatchIngestionWithOverflow(
         const batches = []
         const seenIds: Set<string> = new Set()
         let currentBatch: KafkaMessage[] = []
+
         for (const message of kafkaMessages) {
             const pluginEvent = formPipelineEvent(message)
             const seenKey = `${pluginEvent.team_id}:${pluginEvent.distinct_id}`
-            if (currentBatch.length === batchSize || seenIds.has(seenKey)) {
+
+            // Events with a null key should have been produced to the the
+            // KAFKA_EVENTS_PLUGIN_INGESTION_OVERFLOW topic, so we shouldn't see them here as this consumer's
+            // topic is set to KAFKA_EVENTS_PLUGIN_INGESTION. However, there could be some lingering events
+            // from before the new *_OVERFLOW topic was initialized. Any events with a null key or that
+            // exceed capacity are redirected to the *_OVERFLOW topic.
+            if (message.key == null || ConfiguredLimiter.consume(seenKey, 1) === false) {
+                // Set message key to be null so we know to send it to overflow topic.
+                // We don't want to do it here to preserve the kafka offset handling
+                message.key = null
+            }
+
+            if (currentBatch.length === batchSize || (message.key != null && seenIds.has(seenKey))) {
```

Comment:
could you still make it be just to be safe for the future
```suggestion
            if (currentBatch.length >= batchSize || (message.key != null && seenIds.has(seenKey))) {
```